### PR TITLE
[DROOLS-944] fix FieldReaders serialization

### PR DESCRIPTION
**Pull request by mariofusco**

### CODE REVIEW

File: drools-compiler/src/main/java/org/drools/compiler/rule/builder/QueryElementBuilder.java [hunk 1]
```diff
@@ -499,11 +496,7 @@ public static boolean isVariable( String str ) {
             }
         }
 
-        if ( str.endsWith( ".class" ) ) {
-            return false;
-        }
-
-        return true;
+        return !str.endsWith( ".class" );
```
Comment (psiroky) [inline drools-compiler/src/main/java/org/drools/compiler/rule/builder/QueryElementBuilder.java:499]: This is shorter, but imo bit harder to read :) (but I am fine with leaving it as is)

File: drools-compiler/src/test/java/org/drools/compiler/integrationtests/Misc2Test.java [hunk 2]
```diff
@@ -8205,4 +8205,55 @@ public void testCompilationFailureWithNonExistingField() {
...
+        private Map<String, Object> objects = new HashMap<String, Object>();
+
+        public Map<String, Object> getObjects() {
+            return objects;
+        }
+
+        public void setObjects(Map<String, Object> objects) {
+            this.objects = objects;
+        }
+    }
```
Comment (psiroky) [inline drools-compiler/src/test/java/org/drools/compiler/integrationtests/Misc2Test.java:8211]: I am missing an assert in this test. I know the JIRA is about NPE during de-serialization, but still, could we add an assert that would verify the kbase is the same before and after the serialize+de-serialize calls?

File: drools-core/src/main/java/org/drools/core/base/BaseClassFieldReader.java [hunk 3]
```diff
@@ -184,4 +190,22 @@ public int getHashCode(Object object) {
...
+                                                    ClassNotFoundException {
+        index = in.readInt();
+        valueType = (ValueType) in.readObject();
+        String clsName = in.readUTF();
+
+        try {
+            fieldType = in instanceof DroolsObjectInput ?
+                        Class.forName( clsName, false, ( (DroolsObjectInput) in ).getClassLoader() ) :
+                        Class.forName( clsName );
+        } catch (ClassNotFoundException e) { }
```
Comment (psiroky) [inline drools-core/src/main/java/org/drools/core/base/BaseClassFieldReader.java]: Could you please describe why we don't care about the CNFE? I am sure there is a good reason and I would like to know as my knowledge of Drools is still very limited...

File: drools-core/src/main/java/org/drools/core/base/BaseClassFieldReader.java [hunk 4]
```diff
@@ -184,4 +190,24 @@ public int getHashCode(Object object) {
...
+        index = in.readInt();
+        valueType = (ValueType) in.readObject();
+        String clsName = in.readUTF();
+
+        try {
+            fieldType = in instanceof DroolsObjectInput ?
+                        ClassUtils.getClassFromName( clsName, false, ( (DroolsObjectInput) in ).getClassLoader() ) :
+                        ClassUtils.getClassFromName( clsName );
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException( e );
```
Comment (psiroky) [inline drools-core/src/main/java/org/drools/core/base/BaseClassFieldReader.java:210]: Is there more info we could provide in the exception message? I mean what could be or is the reason the CNFE was thrown, e.g. de-serializing fact which is not on current classpath (just an example, I am not exactly sure what is the cause in this case).

**Comment by mariofusco** [comment]
Merged by https://github.com/droolsjbpm/drools/commit/f0b787326

**Pull request closed without merging**